File: app/src/main/java/com/sn/videoplayer/VideoActivity.java
package com.sn.videoplayer;

import android.os.Bundle;
import android.os.Environment;
import android.util.Log;
import android.view.SurfaceView;
import android.view.View;
import android.widget.Button;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;


import com.sn.plugin_videoplayer.Frame;
import com.sn.plugin_videoplayer.decoder.AudioDecoder;
import com.sn.plugin_videoplayer.decoder.BaseDecoder;
import com.sn.plugin_videoplayer.decoder.IDecoderStateListener;
import com.sn.plugin_videoplayer.decoder.VideoDecoder;

import java.io.File;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class VideoActivity extends AppCompatActivity implements View.OnClickListener, IDecoderStateListener {

    private SurfaceView sfv;

    public static final String TAG = "VideoActivity";

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_video);
        initView();
        initPlayer();
       // initThread();
    }

    private void initThread() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                Log.e(TAG, "开始run" + "");
                while (true) {
                    Log.e(TAG, System.currentTimeMillis() + "");
                    //waitDecode();
                    Log.e(TAG, "开始解锁" + "");
                    try {
                        Thread.sleep(2000);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        }).start();
    }

    private Button videoStart, videoStop, videoRest;

    private void initView() {
        sfv = (SurfaceView) findViewById(R.id.sfv);
        videoStart = (Button) findViewById(R.id.video_start);
        videoStop = (Button) findViewById(R.id.video_stop);
        videoRest = (Button) findViewById(R.id.video_rest);
        videoStart.setOnClickListener(this);
        videoStop.setOnClickListener(this);
        videoRest.setOnClickListener(this);
    }

    private VideoDecoder videoDecoder;
    private AudioDecoder audioDecoder;

    private void initPlayer() {
        String path = Environment.getExternalStorageDirectory().getAbsolutePath() + "/lake.mp4";
        File file = new File(path);
//        if (!file.exists()) {
////            file.mkdir();
//        }
        //创建线程池
        ExecutorService threadPool = Executors.newFixedThreadPool(2);

        //创建视频解码器
        videoDecoder = new VideoDecoder(path, sfv, null);
        threadPool.execute(videoDecoder);
        //创建音频解码器
        audioDecoder = new AudioDecoder(path);
        threadPool.execute(audioDecoder);


        audioDecoder.setStateListener(this);
        videoDecoder.setStateListener(this);

//
//        MP4Repack repack = new MP4Repack(path);
//        repack.start();

    }

    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.video_start) {
            //开启播放
            videoDecoder.goOn();
            audioDecoder.goOn();
        } else if (id == R.id.video_stop) {
            videoDecoder.stop();
            audioDecoder.stop();
        } else if (id == R.id.video_rest) {

        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        videoDecoder.release();
        audioDecoder.release();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        videoDecoder.release();
        audioDecoder.release();
    }

    @Override
    public void decoderPrepare(BaseDecoder decodeJob) {

    }

    @Override
    public void decoderReady(BaseDecoder decodeJob) {

    }

    @Override
    public void decoderRunning(BaseDecoder decodeJob) {

    }

    @Override
    public void decoderPause(BaseDecoder decodeJob) {

    }

    @Override
    public void decodeOneFrame(BaseDecoder decodeJob, Frame frame) {

    }

    @Override
    public void decoderFinish(BaseDecoder decodeJob) {

    }

    @Override
    public void decoderDestroy(BaseDecoder decodeJob) {

    }

    @Override
    public void decoderError(BaseDecoder decodeJob, String msg) {

    }
}

File: app/src/main/java/com/sn/videoplayer/VideoActivity.kt
package com.sn.videoplayer

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.SurfaceHolder
import android.view.SurfaceView
import android.widget.SeekBar
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkInfo
import androidx.work.WorkManager
import androidx.work.workDataOf
import com.sn.videoplayer.data.Config
import com.sn.videoplayer.ffmpeg.demo.*
import com.sn.videoplayer.setting.VideoSettingActivity
import com.sn.videoplayer.view.VideoPlaySurfaceView
import com.sn.videoplayer.worker.CopyFileWork
import kotlinx.android.synthetic.main.activity_video.*
import kotlinx.android.synthetic.main.activity_video.view.*
import org.jetbrains.anko.startActivity

class VideoActivity : AppCompatActivity() {
    var defaultTime = 0
    var seekStatus = false
    var videoFilePath = ""
    private lateinit var  videoPlayView : VideoPlaySurfaceView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_video)
        initData()
        initView()

        DemoNativeInterface.mediaInfoCallBack = (object : MediaInfoCallBack {
            override fun mediaInfoCallBack(mediaInfo: MediaInfo) {
                runOnUiThread {
                    video_seek_bar.setTotalTime(mediaInfo.duration)
                }
            }
        })

        DemoNativeInterface.playerInfoCallBack = (object : PlayerInfoCallBack {
            override fun playerInfo(string: String) {
                runOnUiThread {
                    textView.text = textView.text.toString() + string + "\n"
                }
            }
        })

        DemoNativeInterface.mPlayProgress = (object : PlayProgress {
            override fun progress(progress: Double) {
                if (seekStatus) {
                    return
                }
                runOnUiThread {
                    video_seek_bar.setCurrentTime(progress.toInt())
                }
            }

        })
        var initVideoPlayer = 0;
//        sfv!!.holder!!.addCallback(object : SurfaceHolder.Callback {
//            override fun surfaceChanged(
//                holder: SurfaceHolder,
//                format: Int,
//                width: Int,
//                height: Int
//            ) {
//            }
//
//            override fun surfaceDestroyed(holder: SurfaceHolder) {
//            }
//
//            override fun surfaceCreated(holder: SurfaceHolder) {
//                initVideoPlayer =
//                    fFmpegPlayer.initVideoPlayer(holder.surface, Config.PLAY_FILE_PATH)
//                Log.e(TAG, "initVideoPlayer : $initVideoPlayer")
//                Thread(Runnable {
//                    runOnUiThread {
////                        var videoTime = fFmpegPlayer.getVideoTime(initVideoPlayer)
////                        var videoTime = fFmpegPlayer.getVideoTotalTime(initVideoPlayer)
////                        Log.e(TAG, "videoTime : $videoTime")
////                        seekBar.setTotalTime(videoTime)
//                    }
//                }).start()
//            }
//        })
////
//
//
        video_start.setOnClickListener {
            videoPlayView.start()
        }

        video_stop.setOnClickListener {
            videoPlayView.stop()
        }
        video_seek_after.setOnClickListener {
        }

        video_seek_bar.videoSeekBar.setOnSeekBarChangeListener(object :
            SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
                defaultTime = progress
                video_seek_bar.setCurrentTime(progress.toInt())
                Log.d(TAG, "progress $progress")
            }

            override fun onStartTrackingTouch(seekBar: SeekBar?) {
                seekStatus = true
                Log.d(TAG, "progress onStartTrackingTouch")
            }

            override fun onStopTrackingTouch(seekBar: SeekBar?) {
                DemoNativeInterface.setSeekTime(initVideoPlayer, defaultTime)
                seekStatus = false
                Log.d(TAG, "progress onStopTrackingTouch")
            }

        })
    }

    private fun initData() {
        videoFilePath =  intent.getStringExtra("videoFilePath")!!
    }

    private fun initView() {
        videoPlayView = findViewById(R.id.sfv)
        videoPlayView.setFile(videoFilePath)
        setSupportActionBar(toolbar)
        toolbar.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.video_setting -> {
                    //https://www.jianshu.com/p/4a3ca2c315f2
                    startActivity<VideoSettingActivity>()
                }
            }
            return@setOnMenuItemClickListener true
        }
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        return true
    }


    companion object {
        const val TAG = "VideoActivity"
    }
}